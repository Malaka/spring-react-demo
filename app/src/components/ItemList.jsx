import React, {Component} from 'react';
import {Button, ButtonGroup, Container, Table} from 'reactstrap';
import {Link} from 'react-router-dom';

class ItemList extends Component {

    constructor(props) {
        super(props);
        this.state = {items: [], isLoading: true};
    }

    async componentDidMount() {
        this.setState({isLoading: true});
        await fetch('api/items')
            .then(response => response.json())
            .then(data => this.setState({items: data, isLoading: false}));
    };

    remove = async (id) => {
        await fetch(`/api/item/${id}`, {
            method: 'DELETE',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        }).then(() => {
            let updatedItems = [...this.state.items].filter(i => i.id !== id);
            this.setState({items: updatedItems});
        });
    };

    render() {
        const {items, isLoading} = this.state;

        if (isLoading) {
            return <p>Loading...</p>;
        }

        const itemList = items.map(item => {
            return (
                <tr key={item.id}>
                    <td style={{whiteSpace: 'nowrap'}}>{item.name}</td>
                    <td>{item.description}</td>
                    <td>{item.category && <Button size="sm" tag={Link}
                                                  to={`/search/${item.category.name}`}>{item.category.name}</Button>}</td>
                    <td>
                        <ButtonGroup>
                            <Button size="sm" color="primary" tag={Link} to={"/items/" + item.id}>Edit</Button>
                            <Button size="sm" color="danger" onClick={() => this.remove(item.id)}>Delete</Button>
                        </ButtonGroup>
                    </td>
                </tr>
            );
        });

        return (
            <div>
                <Container fluid>
                    <div className="float-right">
                        <Button color="success" tag={Link} to="/items/new">Add Grocery Item</Button>
                    </div>
                    <h3>My Grocery List</h3>
                    <Table className="mt-4">
                        <thead>
                        <tr>
                            <th width="20%">Name</th>
                            <th>Description</th>
                            <th>Category</th>
                            <th width="10%">Actions</th>
                        </tr>
                        </thead>
                        <tbody>
                        {itemList}
                        </tbody>
                    </Table>
                </Container>
            </div>
        );
    }
}

export default ItemList;