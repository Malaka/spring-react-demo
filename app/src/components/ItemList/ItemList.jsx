import React, {Component} from 'react';
import {Button} from 'reactstrap';
import {Link} from 'react-router-dom';
import {ItemsDisplay} from '../ItemsDisplay/ItemsDisplay'

class ItemList extends Component {

    constructor(props) {
        super(props);
        this.state = {items: [], isLoading: true};
    }

    async componentDidMount() {
        this.setState({isLoading: true});
        await fetch('api/items')
            .then(response => response.json())
            .then(data => this.setState({items: data, isLoading: false}));
    };

    removeHandler = async (id) => {
        await fetch(`/api/item/${id}`, {
            method: 'DELETE',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        }).then(() => {
            let updatedItems = [...this.state.items].filter(i => i.id !== id);
            this.setState({items: updatedItems});
        });
    };

    render() {
        const {items, isLoading} = this.state;

        if (isLoading) {
            return <p>Loading...</p>;
        }

        return (
            <ItemsDisplay items={items} removeHandler={this.removeHandler}>
                <div className="float-right">
                    <Button color="success" tag={Link} to="/items/new">Add Grocery Item</Button>
                </div>
                <h3>My Grocery List</h3>
            </ItemsDisplay>
        );
    }
}

export default ItemList;