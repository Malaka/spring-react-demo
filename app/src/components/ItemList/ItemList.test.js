import * as React from "react";
import {shallow} from "enzyme";
import ItemList from "./ItemList";

describe('ItemList', () => {
    it('fetches data from server when server returns a successful response', async () => {
        const mockSuccessResponse = [{
            "id": 6,
            "name": "Fresh Milk",
            "description": "This is fresh",
            "category": {"id": 1, "name": "Milk"}
        }];
        const mockJsonPromise = Promise.resolve(mockSuccessResponse);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

        const wrapper = shallow(<ItemList/>);

        expect(global.fetch).toHaveBeenCalledTimes(1);
        expect(global.fetch).toHaveBeenCalledWith('api/items');

        await flushPromises();
        expect(wrapper.state()).toEqual({
                "isLoading": false,
                "items": [{
                    "category": {"id": 1, "name": "Milk"},
                    "description": "This is fresh",
                    "id": 6,
                    "name": "Fresh Milk"
                }]
            }
        );
        expect(wrapper).toMatchSnapshot();
        global.fetch.mockClear();
    });
});

const flushPromises = () => {
    return new Promise(resolve => setImmediate(resolve));
};