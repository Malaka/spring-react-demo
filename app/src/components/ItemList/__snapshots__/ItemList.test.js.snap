// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList fetches data from server when server returns a successful response 1`] = `
<ItemsDisplay
  items={
    Array [
      Object {
        "category": Object {
          "id": 1,
          "name": "Milk",
        },
        "description": "This is fresh",
        "id": 6,
        "name": "Fresh Milk",
      },
    ]
  }
  removeHandler={[Function]}
>
  <div
    className="float-right"
  >
    <Button
      color="success"
      tag={[Function]}
      to="/items/new"
    >
      Add Grocery Item
    </Button>
  </div>
  <h3>
    My Grocery List
  </h3>
</ItemsDisplay>
`;
